test/widgets/action_sheet_test.dart

```diff
@@ -1188,6 +1189,199 @@ void main() {
+    }
+
+    group('showChannelActionSheet', () {
+      void checkButtons() {
+        check(actionSheetFinder).findsOne();
+        checkButton('Mark channel as read');
+      }
+
+      testWidgets('show from inbox', (tester) async {
+        await prepare();
+        await showFromInbox(tester);
+        checkButtons();
+      });
+
+      testWidgets('show from subscription list', (tester) async {
+        await prepare();
+        await showFromSubscriptionList(tester);
+        checkButtons();
+      });
+
+      testWidgets('show with no unread messages', (tester) async {
+        await prepare(unreadMsgs: eg.unreadMsgs());
+        await showFromSubscriptionList(tester);
+        check(actionSheetFinder).findsNothing();
+      });
+
+      testWidgets('show from app bar in channel narrow', (tester) async {
+        await prepare();
+        await showFromAppBar(tester);
+        checkButtons();
+      });
+
+      testWidgets('show from app bar in topic narrow', (tester) async {
+        await prepare();
+        await showFromAppBar(tester, isTopicNarrow: true);
+        checkButtons();
+      });
+
+      testWidgets('show from recipient header', (tester) async {
+        await prepare();
+        await showFromRecipientHeader(tester);
+        checkButtons();
+      });
+    });
+
+    group('MarkChannelAsReadButton', () {
+      void checkRequest(int streamId) {
+        check(connection.takeRequests()).single.isA<http.Request>()
+          ..method.equals('POST')
+          ..url.path.equals('/api/v1/messages/flags/narrow')
+          ..bodyFields.deepEquals({
+            'anchor': 'oldest',
+            'include_anchor': 'false',
+            'num_before': '0',
+            'num_after': '1000',
+            'narrow': jsonEncode([
+              {'operator': 'stream', 'operand': streamId},
+              {'operator': 'is', 'operand': 'unread'}
+            ]),
+            'op': 'add',
+            'flag': 'read',
+          });
+      }
+
+      testWidgets('happy path from inbox', (tester) async {
+        await prepare();
+        await showFromInbox(tester);
+        connection.prepare(json: UpdateMessageFlagsForNarrowResult(
+          processedCount: 1, updatedCount: 1,
+          firstProcessedId: null, lastProcessedId: null,
+          foundOldest: true, foundNewest: true).toJson());
+        await tester.tap(findButtonForLabel('Mark channel as read'));
+        await tester.pumpAndSettle();
+
+        checkNoErrorDialog(tester);
+        checkRequest(someChannel.streamId);
+      });
+
+      testWidgets('request fails', (tester) async {
+        await prepare();
+        await showFromInbox(tester);
+
+        connection.prepare(httpException: http.ClientException('Oops'));
+
+        await tester.tap(findButtonForLabel('Mark channel as read'));
+        await tester.pumpAndSettle();
+
+        checkErrorDialog(tester,
+          expectedTitle: "Mark as read failed");
+        checkRequest(someChannel.streamId);
```

Comment:
nit: the request logically happens before the error dialog, so put checks in the same order